File: leetcode/unique-paths_topDown.cpp
class Solution {
public:
    bool segCheck(int i,int j,int m, int n){
        return (0<=i && i<m && 0<=j&& j<n);
    }
    int robotMover(vector<vector<int>>&dp,int m, int n,int curI,int curJ){
        if(!segCheck(curI,curJ,m,n)) return 0;
        else{
            if( dp[curI][curJ]!=0) return dp[curI][curJ];
            else
                return dp[curI][curJ]=robotMover(dp,m,n,curI-1, curJ)+robotMover(dp,m,n,curI, curJ-1);
        }
    }
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m,vector<int>(n,0));
        dp[0][0]=1;
        robotMover(dp,m,n,m-1,n-1);
        return dp[m-1][n-1];
    }
};

File: leetcode/uniquePaths_topDown.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp=[[0]*n for i in range(m)]
        dp[0][0]=1
        def topDownDp(dp:list,index:int,jndex:int)->int:
            if not (0<=index<m and 0<=jndex<n): return 0
            if dp[index][jndex]!=0: return dp[index][jndex]
            dp[index][jndex]=topDownDp(dp,index-1,jndex)+topDownDp(dp,index,jndex-1)
            return dp[index][jndex]
        return topDownDp(dp,m-1,n-1)
